crates/turborepo-scm/src/git.rs

```diff
@@ -94,11 +137,72 @@ impl Git {
         Ok(output.trim().to_owned())
     }
 
+    /// for GitHub Actions environment variables, see: https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/store-information-in-variables#default-environment-variables
+    pub fn get_github_base_ref() -> Option<String> {
+        // make sure we're running in a CI environment
+        if env::var("CI").ok()?.as_str() != "true" {
+            return None;
+        }
+
+        // make sure we're running in a GitHub Action
+        if env::var("GITHUB_ACTIONS").ok()?.as_str() != "true" {
+            return None;
+        }
+
+        /*
+         * The name of the base ref or target branch of the pull request in a
+         * workflow run.
+         *
+         * This variable only has a value when the event that triggers a workflow run
+         * is either `pull_request` or `pull_request_target`.
+         * For example, `main`
+         *
+         * So environment variable is empty in a regular commit
+         */
+        if let Ok(pr) = env::var("GITHUB_BASE_REF") {
+            if !pr.is_empty() {
+                return Some(pr);
+            }
+        }
+
+        // we must be in a push event
+        // try reading from the GITHUB_EVENT_PATH file
+        if let Ok(event_path) = env::var("GITHUB_EVENT_PATH") {
+            // Try to open the event file and read the contents
+            let mut file = File::open(event_path).ok()?;
+            let mut data = String::new();
+            file.read_to_string(&mut data).ok()?;
+
+            // Parse the JSON data from the file
+            let json: GitHubEvent = serde_json::from_str(&data).ok()?;
+
+            // Extract the base ref from the pull request event if available
+            let base_ref = &json.before;
+
+            // the base_ref will be UNKNOWN_SHA on first push to a new branch and force
+            // pushes
+            if base_ref == UNKNOWN_SHA {
+                return json.get_first_commit_id();
+            }
+
+            if base_ref.is_empty() {
+                return None;
+            }
+
+            return Some(base_ref.to_string());
+        }
+        None
+    }
+
     fn resolve_base<'a>(&self, base_override: Option<&'a str>) -> Result<&'a str, Error> {
         if let Some(valid_from) = base_override {
             return Ok(valid_from);
         }
 
+        if let Some(github_base_ref) = Self::get_github_base_ref() {
+            return Ok(Box::leak(github_base_ref.into_boxed_str()));
```

Comment:
What was the reason for this? Leaking memory is probably not ideal, even if it's a small amount. Could we change this function to return a `String`? If we want to be efficient a `Cow<'a, str>` could work, but that's probably overkill

Reply:
I went in circles and circles on this.  I'm not sure what the problem is, but I couldn't find anything else that could work.  Do you know how this could be rewritten, concretely, that would work.

I couldn't believe this was hard because it seemed like it should be easy, but there's something I'm missing.

Reply:
update: found a way to do it 4d138792ffc65370a9af9c75a8891d35503d716f like you suggested